### PR TITLE
Fix appending issues

### LINKED ISSUE #82: Array of strings handling not working
Separated this issue from #80 to create smaller issues. 

Fix is not able to operate with array of strings, It cannot create new ones and cannot refrence existing once.

It should be able to refrence a string/value with e.g.: `test.$append`, `test.*` See: http://librecat.org/Catmandu/#wildcards
It would operate on them as items. (This could also be true for handling entire objects as items.)

[Broken workflow](
https://test.metafacture.org/playground/?data=%7B%0A++++%22arrayOfStringimals%22%3A+%5B%22cat%22%2C%22dog%22%2C%22fox%22%5D%0A%7D&flux=PG_DATA%0A%7C+as-records%0A%7C+decode-json%0A%7C+fix%0A%7C+encode-json%28prettyPrinting%3D%22true%22%29%0A%7C+print%0A%3B&fix=add_field%28%22arrayOfStringimals%5B%5D.%24append%22%2C%22duck%22%29&active-editor=fix)

It works with a subfield: [here](https://test.metafacture.org/playground/?data=%7B%0A++++%22arrayOfStringimals%22%3A+%5B%22cat%22%2C%22dog%22%2C%22fox%22%5D%0A%7D&flux=PG_DATA%0A%7C+as-records%0A%7C+decode-json%0A%7C+fix%0A%7C+encode-json%28prettyPrinting%3D%22true%22%29%0A%7C+print%0A%3B&fix=add_field%28%22arrayOfStringimals%5B%5D.%24append.test%22%2C%22duck%22%29&active-editor=fix)

It could be that this issue is too connected to #80 then we should close it.

### PR DESCRIPTION
**Pull request by fsteeg**
Will resolve #82 and #80.

### CODE REVIEW

File: metafix/src/main/java/org/metafacture/metafix/Value.java [hunk 1]
```diff
@@ -264,6 +264,23 @@ void apply(final Hash hash, final String field, final String value) {
                 void apply(final Hash hash, final String field, final String value) {
                     hash.add(field, new Value(value));
                 }
+            },
+            /* For an indexed representation of arrays as hashes with 1, 2, 3 etc. keys.
+             * i.e. ["a", "b", "c"] as { "1":"a", "2":"b", "3": "c" }
+             * This is what is produced by JsonDecoder and Metafix itself for arrays.
```
Comment (blackwinter) [inline metafix/src/main/java/org/metafacture/metafix/Value.java:270]: Where does Metafix do this?

File: metafix/src/main/java/org/metafacture/metafix/Value.java [hunk 2]
```diff
@@ -264,6 +264,23 @@ void apply(final Hash hash, final String field, final String value) {
...
+             * TODO? maybe this would be a good general internal representation, resulting
+             * in every value being either a hash or a string, no more separate array type.*/
+            INDEXED {
+                @Override
+                void apply(final Hash hash, final String field, final String value) {
+                    final Value newHash = newHash();
+                    newHash.asHash().put(field, new Value(value));
+                    hash.add(nextIndex(hash), newHash);
```
Comment (blackwinter) [inline metafix/src/main/java/org/metafacture/metafix/Value.java:278]:
```suggestion
                    hash.add(nextIndex(hash), newHash(h -> h.put(field, new Value(value))));
```

File: metafix/src/test/java/org/metafacture/metafix/MetafixRecordTest.java [hunk 3]
```diff
@@ -264,6 +264,113 @@ public void add() {
             });
     }
 
+    @Test
+    public void addWithAppendInArray() {
+        MetafixTestHelpers.assertFix(streamReceiver, Arrays.asList(
+                "add_field('names.$append','patrick')",
+                "vacuum()"),
```
Comment (blackwinter) [inline metafix/src/test/java/org/metafacture/metafix/MetafixRecordTest.java:271]: Why `vacuum()`? (Also in following tests.)

**Review by blackwinter — APPROVED** [review]
Aside from explicit array marking and still somewhat confusing list semantics (merging), :+1:

File: metafix/src/main/java/org/metafacture/metafix/Value.java (hunk 1 above)
Comment (fsteeg) [reply metafix/src/main/java/org/metafacture/metafix/Value.java:270]: I meant the general behavior for repeated input fields, which become a list internally, and are emitted in that indexed form (see e.g. in [this test](https://github.com/metafacture/metafacture-fix/blob/4e26885f346b4cc3542c716661b4c3084ae12bae/metafix/src/test/java/org/metafacture/metafix/MetafixMethodTest.java#L54)).

**Comment by fsteeg** [comment]
Thanks for the review! Addressed your feedback in https://github.com/metafacture/metafacture-fix/pull/87/commits/adb06930f92d1dae241344dc4f83087b6e75d778.

File: metafix/src/main/java/org/metafacture/metafix/Value.java (hunk 1 above)
Comment (blackwinter) [reply metafix/src/main/java/org/metafacture/metafix/Value.java:270]: I see, thanks for the clarification.

**Comment by blackwinter** [comment]
Test `complexAppendWithArrayOfObjects()` still fails:

```
org.mockito.exceptions.verification.VerificationInOrderFailure: 
Verification in order failure
Wanted but not invoked:
streamReceiver.startEntity("4");
-> at org.metafacture.metafix.MetafixRecordTest.lambda$complexAppendWithArrayOfObjects$31(MetafixRecordTest.java:466)
Wanted anywhere AFTER following interaction:
streamReceiver.endEntity();
```

Fix string:

```
copy_field('others', 'animals[].$append')
move_field('fictional', 'animals[].$append')
add_field('animals[].$append.animal', 'earthworm')
```

```
[Mockito] Interactions of: streamReceiver
     1. streamReceiver.startRecord("1");
      -> at org.metafacture.metafix.Metafix.endRecord(Metafix.java:146)
     2. streamReceiver.startEntity("animals[]");
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:164)
     3. streamReceiver.startEntity("1");
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:164)
     4. streamReceiver.literal("animal", "dog");
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:169)
     5. streamReceiver.endEntity();
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:166)
     6. streamReceiver.startEntity("2");
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:164)
     7. streamReceiver.literal("animal", "cat");
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:169)
     8. streamReceiver.endEntity();
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:166)
     9. streamReceiver.startEntity("3");
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:164)
     10. streamReceiver.literal(
        "animal",
        "earthworm"
    );
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:169)
     11. streamReceiver.endEntity();
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:166)
     12. streamReceiver.endEntity();
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:166)
     13. streamReceiver.startEntity("others");
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:164)
     14. streamReceiver.literal("animal", "human");
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:169)
     15. streamReceiver.endEntity();
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:166)
     16. streamReceiver.endRecord();
      -> at org.metafacture.metafix.Metafix.endRecord(Metafix.java:149)
```

Is the test wrong?

**Comment by fsteeg** [comment]
> Is the test wrong?

No, it does not work yet for (indexed) arrays of objects (only strings).

To keep changes small and focused, I'd say we disable the test and work on that in #92.

**Review by blackwinter — APPROVED** [review]
> To keep changes small and focused, I'd say we disable the test and work on that in #92.

Okay.

**Pull request merged**